### PR TITLE
[dv/clkmgr] Add exclusions and coverage pragmas

### PR DESCRIPTION
**Pull request by matutem**
To get coverage back above 90% for all metrics this adds some UNR and manual exclusions (for cases UNR fails to detect). Add some VCS coverage pragmas to turn off coverage for non-synthesized code.

Signed-off-by: Guillermo Maturana <maturana@google.com>

### CODE REVIEW

File: hw/dv/sv/dv_base_reg/dv_base_reg_pkg.sv [hunk 1]
```diff
@@ -68,9 +68,9 @@ package dv_base_reg_pkg;
     string        msg_id = {dv_base_reg_pkg::msg_id, "::decode_csr_or_field"};
 
     if ($cast(csr, ptr)) begin
-      // return csr object with null field; set the mask to all 1s and shift to 0
+      // return csr object with null field; set the mask to the width's all 1s and shift to 0
       result.csr = csr;
-      result.mask = '1;
+      result.mask = (1 << csr.get_n_bits()) - 1;
```
Comment (matutem) [inline hw/dv/sv/dv_base_reg/dv_base_reg_pkg.sv:73]: This is unrelated to coverage.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// This contains some obvious exclusions the UNR tool didn't flag.
+
+//==================================================
+// This file contains the Excluded objects
+// Generated By User: maturana
+// Format Version: 2
+// Date: Thu Sep 29 13:57:41 2022
+// ExclMode: default
+//==================================================
+CHECKSUM: "2301929872 1660332954"
```
Comment (matutem) [inline hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: Indeed, these exclusions were not caught by the UNR tool.

File: hw/ip/prim/rtl/prim_pulse_sync.sv [hunk 3]
```diff
@@ -36,6 +36,7 @@ module prim_pulse_sync (
   // source active must come far enough such that the destination domain has time
   // to create a valid pulse.
 `ifdef INC_ASSERT
+//VCS coverage off
```
Comment (matutem) [inline hw/ip/prim/rtl/prim_pulse_sync.sv:39]: Here and below we just explicitly disable coverage through the line with coverage on.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 2 above)
Comment (cindychip) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: I believe I saw the same issue in kmac regarding the step_i :)

**Review by cindychip — COMMENTED** [review]
LGTM, would refer to the designer to approve the UNR

File: hw/ip/prim/rtl/prim_pulse_sync.sv (hunk 3 above)
Comment (sriyerg) [reply hw/ip/prim/rtl/prim_pulse_sync.sv:39]: Could we make make this tool-agnostic (I am sure most tools support such inline pragmas). Could we define something like this in our `dv_macros.svh`?

```systemverilog
`ifdef VCS
  `define PRAGMA_COVERAGE_ON /``/VCS coverage on
  `define PRAGMA_COVERAGE_ON /``/VCS coverage off
`endif
```

File: hw/dv/sv/dv_base_reg/dv_base_reg_pkg.sv (hunk 1 above)
Comment (sriyerg) [inline hw/dv/sv/dv_base_reg/dv_base_reg_pkg.sv:73]: If `csr.get_n_bits()` = 32, then won't this overflow? I think `result.mask` is 32 bits.

Comment (sriyerg) [reply hw/dv/sv/dv_base_reg/dv_base_reg_pkg.sv:73]: Hmm, actually `result.mask` will be all 1s, making it equivalent to the previous code for `csr.get_n_bits()` == 32 case. For `csr.get_n_bits()` < 32, this would be an improvement. LGTM.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 2 above)
Comment (sriyerg) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: Should we have an AE investigate these?

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+INSTANCE: tb.dut.u_clk_main_aes_trans.u_idle_cnt
+ANNOTATION: "[UNSUPPORTED] Input tied to a constant, and unr doesn't detect it."
```
Comment (sriyerg) [inline hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:16]:
```suggestion
ANNOTATION: "[UNR] Input tied to a constant."
```

Aren't these UNR?

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 2 above)
Comment (sriyerg) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: Is this due to an input port being tied off? That would explain why the UNR tool is not able to find these, because it does not know we tie these off in our testbench. I'm sure there must be a way to feed that information to the tool.

Comment (matutem) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: This is an input to an internal instance, and it is driven within the design, so it is just an unfortunate tool deficiency that was reported, and I think accepted as a bug.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 4 above)
Comment (matutem) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:16]: I see... I had it as UNR in an ancient PR that tried this, but then I saw some exclusion file stating UNSUPPORTED for a case like this and I assumed the writer was better informed...I assume an annotation change doesn't invalidate the checksum, since it is not in the RTL. Will update it.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 2 above)
Comment (sriyerg) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: Ok, got it.

File: hw/ip/prim/rtl/prim_pulse_sync.sv (hunk 3 above)
Comment (matutem) [reply hw/ip/prim/rtl/prim_pulse_sync.sv:39]: I will do it in a separate PR, since there are a few other files related to this.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 2 above)
Comment (tjaychen) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el:14]: these look good to me!

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+// This file contains the Excluded objects
+// Generated By User: maturana
+// Format Version: 2
+// Date: Thu Oct  6 16:19:29 2022
+// ExclMode: default
+//==================================================
+CHECKSUM: "1607649888 3304441462"
+INSTANCE: tb.dut.u_reg.u_io_meas_ctrl_en_cdc.u_arb
+ANNOTATION: "VC_COV_UNR"
+Condition 5 "593451913" "(((!gen_wr_req.dst_req)) && gen_wr_req.dst_lat_d) 1 -1" (1 "01")
```
Comment (tjaychen) [inline hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:18]: i really can't tell which combination of values they think is unreachable here...

Comment (tjaychen) [reply hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:18]: ah okay i think i finally figured it out from the coverage report. This line to line 35 are okay.

File: hw/ip/prim/rtl/prim_pulse_sync.sv (hunk 3 above)
Comment (matutem) [reply hw/ip/prim/rtl/prim_pulse_sync.sv:39]: Well, verible cannot handle these kinds of macros: see https://github.com/lowRISC/opentitan/issues/15330, so I will just inline the coverage directives until verible is fixed.

File: hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el (hunk 4 above)
Comment (matutem) [reply hw/ip/clkmgr/dv/cov/clkmgr_manual_cov_unr_excl.el]: Done.

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+CHECKSUM: "1607649888 3304441462"
+INSTANCE: tb.dut.u_reg.u_main_meas_ctrl_en_cdc.u_arb
+ANNOTATION: "VC_COV_UNR"
+Condition 5 "593451913" "(((!gen_wr_req.dst_req)) && gen_wr_req.dst_lat_d) 1 -1" (1 "01")
+CHECKSUM: "1607649888 3304441462"
+INSTANCE: tb.dut.u_reg.u_usb_meas_ctrl_en_cdc.u_arb
+ANNOTATION: "VC_COV_UNR"
+Condition 5 "593451913" "(((!gen_wr_req.dst_req)) && gen_wr_req.dst_lat_d) 1 -1" (1 "01")
+CHECKSUM: "2972535896 3274445021"
+INSTANCE: tb.dut.u_reg.u_clk_hints_status_clk_main_aes_val
```
Comment (matutem) [inline hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:36]: All the wr_en exclusions are because de is always on for these instances since they are sro.

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+ANNOTATION: "VC_COV_UNR"
+Condition 1 "2397158838" "(wr_en ? wr_data : qs) 1 -1" (1 "0")
+CHECKSUM: "3864567832 3302102175"
+INSTANCE: tb.dut.u_main_root_ctrl.u_scanmode_sync
+ANNOTATION: "VC_COV_UNR"
+Block 1 "254557145" "if ((!rst_ni))"
+ANNOTATION: "VC_COV_UNR"
+Block 3 "1282307173" "gen_no_flops.unused_logic <= mubi_i;"
+CHECKSUM: "3864567832 3302102175"
+INSTANCE: tb.dut.u_io_root_ctrl.u_scanmode_sync
```
Comment (matutem) [inline hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:58]: All these "if (!rst_ni)" are for unused logic in prim_mubi4_sync: https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/prim/rtl/prim_mubi4_sync.sv;drc=10daa25ec9d8bebe6d86b6b20694cad1c2a40448;l=143

I will add explicit coverage exclusions in a separate PR, since I will need to change the mubi_sync templates.

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+CHECKSUM: "215202837 3193272610"
+INSTANCE: tb.dut.u_io_meas.u_meas.gen_clk_timeout_chk.u_timeout_ref_to_clk
+ANNOTATION: "VC_COV_UNR"
+Branch 0 "3003057152" "(!rst_ni)" (4) "(!rst_ni) 0,0,0,0"
+CHECKSUM: "215202837 3193272610"
+INSTANCE: tb.dut.u_io_div2_meas.u_meas.gen_clk_timeout_chk.u_timeout_ref_to_clk
+ANNOTATION: "VC_COV_UNR"
+Branch 0 "3003057152" "(!rst_ni)" (4) "(!rst_ni) 0,0,0,0"
+CHECKSUM: "215202837 3193272610"
+INSTANCE: tb.dut.u_io_div4_meas.u_meas.gen_clk_timeout_chk.u_timeout_ref_to_clk
```
Comment (matutem) [inline hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:90]: The u_timeout_ref_to_clk seem to be due to a missing else in the flop at https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/prim/rtl/prim_clock_timeout.sv;drc=df6bbf5f67659efa109d90a1d730a68d35a5a590;l=40

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+ANNOTATION: "VC_COV_UNR"
+Branch 0 "3003057152" "(!rst_ni)" (4) "(!rst_ni) 0,0,0,0"
+CHECKSUM: "215202837 3193272610"
+INSTANCE: tb.dut.u_usb_meas.u_meas.gen_clk_timeout_chk.u_timeout_ref_to_clk
+ANNOTATION: "VC_COV_UNR"
+Branch 0 "3003057152" "(!rst_ni)" (4) "(!rst_ni) 0,0,0,0"
+CHECKSUM: "3864567832 1223333544"
+INSTANCE: tb.dut.u_main_root_ctrl.u_scanmode_sync
+ANNOTATION: "VC_COV_UNR"
+Branch 0 "455461571" "(!rst_ni)" (0) "(!rst_ni) 1"
```
Comment (matutem) [inline hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:104]: It seems these are because the prim_mubi4_syncs with no flops tie rst_ni to a constant. This is non-synthesized code so will be removed in a subsequent PR.

File: hw/ip/prim/rtl/prim_macros.svh [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+// Don't forget to add a closing ON pragma after the code to be skipped.
+//
+// Some notes:
+// - The first line is for VCS, the second for xcelium. It is okay to issue both regardless of
+//   the tool used.
+// - For xcelium it is possible to discriminate between metrics to be disabled as follows
+//   //pragma coverage <metric> = on/off
+//   where metric can be block | expr | toggle | fsm.
+
+// TODO(verible bug) Verible seems to get confused by these macros, so
```
Comment (sriyerg) [inline hw/ip/prim/rtl/prim_macros.svh:17]: Do you want to file an issue with Verible:
https://github.com/chipsalliance/verible/issues

```suggestion
// TODO(chipsalliance/verible#NNN) Verible seems to get confused by these macros, so
```

File: hw/ip/prim/rtl/prim_clock_meas.sv [hunk 11]
```diff
@@ -127,7 +127,11 @@ module prim_clock_meas #(
         end
       end
 
+//VCS coverage off
```
Comment (sriyerg) [inline hw/ip/prim/rtl/prim_clock_meas.sv:130]: Is it problematic to align these comments with the code?

File: hw/ip/prim/rtl/prim_macros.svh [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+/``/VCS coverage off \
+/``/ pragma coverage off
+`endif
+
+`ifndef PRAGMA_COVERAGE_ON
+`define PRAGMA_COVERAGE_ON \
+/``/VCS coverage on \
+/``/ pragma coverage on
+`endif
+*/
```
Comment (sriyerg) [inline hw/ip/prim/rtl/prim_macros.svh:31]: If Verible is the only issue, then we could disable the associated rule that is producing the failure inline here :)

```systemverilog
// verilog_lint: waive-start <rule-X>
`ifndef PRAGMA_COVERAGE_OFF
`define PRAGMA_COVERAGE_OFF \
/``/VCS coverage off \
/``/ pragma coverage off
`endif

`ifndef PRAGMA_COVERAGE_ON
`define PRAGMA_COVERAGE_ON \
/``/VCS coverage on \
/``/ pragma coverage on
`endif
// verilog_lint: waive-stop <rule-X>
```

Comment (sriyerg) [reply hw/ip/prim/rtl/prim_macros.svh:31]: This is documented here: https://github.com/chipsalliance/verible/tree/master/verilog/tools/lint

File: hw/ip/prim/rtl/prim_pulse_sync.sv (hunk 3 above)
Comment (sriyerg) [reply hw/ip/prim/rtl/prim_pulse_sync.sv:39]: Aah - I added a comment on how to circumvent this above.

File: hw/ip/prim/rtl/prim_macros.svh (hunk 10 above)
Comment (matutem) [reply hw/ip/prim/rtl/prim_macros.svh:17]: I filed a bug to the internal verible team... I'll file it on the public side so we can link the bug.

File: hw/ip/prim/rtl/prim_macros.svh (hunk 12 above)
Comment (matutem) [reply hw/ip/prim/rtl/prim_macros.svh:31]: Verible gets a syntax error with these macros, and syntax errors cannot be waived, so I am afraid this is as far as we can go for now.

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el (hunk 8 above)
Comment (tjaychen) [reply hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:90]: but doesn't the missing `else` just mean the count remains the same?

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el (hunk 7 above)
Comment (tjaychen) [reply hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:58]: sounds good, let me know if you need any help here. 
We could alternatively just remove that unused logic and add lint waivers.  That's also an option.

File: hw/ip/prim/rtl/prim_macros.svh (hunk 12 above)
Comment (sriyerg) [reply hw/ip/prim/rtl/prim_macros.svh:31]: Ohh, nevermind then.

File: hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el (hunk 8 above)
Comment (matutem) [reply hw/ip/clkmgr/dv/cov/clkmgr_cov_unr_excl.el:40]: Yes it does, but coverage complains we never visit the "else". It is unreachable probably because !en_i is already dealt with in line 34.

File: hw/ip/prim/rtl/prim_pulse_sync.sv (hunk 3 above)
Comment (matutem) [reply hw/ip/prim/rtl/prim_pulse_sync.sv]: It doesn't work because syntax error cannot be waived.

File: hw/ip/prim/rtl/prim_clock_meas.sv (hunk 11 above)
Comment (matutem) [reply hw/ip/prim/rtl/prim_clock_meas.sv]: Done.

**Review by tjaychen — APPROVED** [review]
looks good from my end.

**Review by sriyerg — APPROVED** [review]
